### DISCUSSION TITLE
False positive "Argument missing for parameter "self"?

### DISCUSSION THREAD

**Discussion by sergeykorablin** [microsoft/pyright#6669] [Q&A] [answered]
I see annoying error message "Argument missing for parameter "self". Is it a bug?

```python
import asyncio
import telegram

bot = telegram.Bot(token="some-token")
async def send_msg():
    await bot.send_message(chat_id="channel-id", text="Hello World")  # Argument missing for parameter "self"

asyncio.run(send_msg())
```
```sh
$ .local/share/nvim/mason/bin/pyright --version
pyright 1.1.339

$ python -V          
Python 3.10.12

$ pip freeze
python-telegram-bot==20.7
```

**Comment by erictraut** [comment, marked as answer]
This is a bug in the `python-telegram-bot` library. The `send_message` function is decorated with `@_log`, and the `_log` function is not properly annotated. If it were completely unannotated, pyright would infer its type correctly, but it is partially annotated, so pyright assumes that it shouldn't override the intent of the author.

The correct way to type `_log` is to use a `TypeVar` bound to a `Callable` or a `Callable` with a `ParamSpec`.

```python
    # Option 1: TypeVar bound to a Callable
    T = TypeVar("T", bound=Callable[..., Any])

    def _log(func: T) -> T: ...

    # Option 2: Callable with a ParamSpec
    P = ParamSpec("P")
    R = TypeVar("R")

    def _log(func: Callable[P, R]) -> Callable[P, R]: ...
```

A third option is to remove all annotations for `_log`. Pyright will then correctly infer the return type. However, this won't work with other type checkers like mypy, which don't support return type inference.

I recommend that you file a bug with the maintainers of this library.

**Reply by sergeykorablin** [reply]
Thanks for the detailed explanation!

**Comment by leakedby** [comment]
file: pyproject.toml

```
[tool.pyright]
venvPath = "./"
venv = ".venv"
reportMissingImports = true
reportMissingTypeStubs = false
reportMissingParameterType = false
typeCheckingMode = "off"
pythonVersion = "3.11.6"
``` 
Turn off this by set **reportMissingParameterType = false**
